### ISSUE TITLE
Compiling under Windows (VS2015)

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#1749]
[Neovim on AppVeyor](https://ci.appveyor.com/project/equalsraf/neovim)

[Comprehensive notes on building for Windows](https://gist.github.com/equalsraf/4685f7aef022a36c26d5)

- [x] Attempt to build neovim using cmake and VS2013
- [x] Build & install libuv somewhere that cmake can find it
- [x] start thinking about how to split platform code (os_unix.c os_msvc.c etc)
- [x] Start splitting compiler options in CMake (#810)
- [x] fix listening server, Windows [pipe](http://msdn.microsoft.com/en-us/library/windows/desktop/aa365150%28v=vs.85%29.aspx) names must always start be in the form `"\\\\.\\pipe\\pipename"` (d52fb89fb85984df23b093637b9a8738fdcfe18a )
- [x] `gendeclarations.lua` patch. https://github.com/neovim/neovim/pull/877
- [x] Port if_cscope to use libuv (https://github.com/neovim/neovim/pull/810#issuecomment-45277957) #3927
- [x] we need a os_file_is_readable() in os/fs.c to fix eval.c:f_file_readable() [#3166](https://github.com/neovim/neovim/pull/3166)
- [ ] check if Neovim can work on windows without libintl (find HAVE_WORKING_LIBINTL, implement ex_cmds2.c:get_locale_val()?). It builds but the old code was removed, and needs testing.
- [ ] Refactor vim_FulllName
    - Possibly addressed by @splinterofchaos pending PR
- [x] Add include guards: sys/time.h unistd.h
- [x] Reorder includes here and there, i.e. ensure platform includes and/or vim.h are included before anything else
- [x] bring back utf16_to_enc()/enc_to_utf16() - needed to convert string encoding when calling windows functions #4837
- [x] declare VS 2015 as minimum MSVC ([comment](https://github.com/neovim/neovim/issues/1749#issuecomment-111690340))
- [ ] busted tests: lubyk/xml#4 should work for MSVC but not for Mingw
- [ ] lua nvim-client does not build on Windows.
- [ ] Get ffi working with MSVC - specially how to handle __pragma (see d530fb62dd52a0724fab38be49514a47464eef75 for more details)
- [ ] `os_setperm()` needs to be re-thought; it is nonsense on Windows [because of `uv_fs_chmod`](https://github.com/libuv/libuv/blob/e76b8838e51f6e8f1944b6c6d50c3948ed764a0b/src/win/fs.c#L1072-L1081).

From https://github.com/neovim/neovim/issues/1749#issuecomment-90766351 :
- [x] terminal emulator and libvterm
    - libvterm builds in MSVC
- [x] pty_process.c seems UNIX specific (termios.h)
    - [For now pipe_process.c can be used(disable pty_process.c compilation and `#define pty_*` functions as `pipe_*`. Later I will try to implement a winpty_process.c on top of winpty](https://github.com/neovim/neovim/issues/1749#issuecomment-91794020)
- [x] job.c, waitpid(-1, ...) equivalent for Windows
    - "SIGCHLD handling can simply be removed(through `#ifdef`?), it only exists to support processes spawned by pty_process."
    - [guarded those blocks with a define FEAT_PTY_PROCESS.](https://github.com/neovim/neovim/issues/1749#issuecomment-91794020)

**Comment by equalsraf** [comment]
@justinmk sure delete away. One extra todo item for the list above

- [ ] fix listening server, Windows [pipe](http://msdn.microsoft.com/en-us/library/windows/desktop/aa365150%28v=vs.85%29.aspx) names must always start be in the form "\\\\.\\pipe\\pipename"

**Comment by equalsraf** [comment]
And another 

- [ ] check if Neovim can work on windows without libintl (find HAVE_WORKING_LIBINTL, implement ex_cmds2.c:get_locale_val()?). It builds but the old code was removed, and needs testing.

**Comment by justinmk** [comment]
updated

**Comment by equalsraf** [comment]
Commit 31c8440fee0bd694093d83af5dee20e3c3dc876c uses SSIZE_T which is POSIX, but does not exist in Windows, looking for suitable replacements (using INT_MAX for now).

**Comment by elmart** [comment]
`intmax_t` is ok, yes. `ptrdiff_t` could be another alternative.
But, to me, the ideal alternative is neither of those.
I don't like C's convention of mixing two concerns (proper result and error indication) in the same variable.
I'd prefer using two variables (one `size_t` variable for the proper result, and one `bool` variable for error indication.
It seems `ssize_t` is only used by `find_term_bykeys`, which is used only once. So, refactoring that function to get rid of `ssize_t` altogether seems to me like the best alternative. You'd have several alternatives for the new function signature:
- `size_t find_term_bykeys(char_u *src, bool *error);`
- `bool find_term_bykeys(char_u *src, size_t *index);`
- `compound_result_T find_term_bykeys(char_u *src);`, where `compound_result_T` (named some other way) would be a typedef for `struct {size_t index; bool error;}`.
This is something I'd do for the entire codebase, little by little. The only drawback is having to define a per-function return type. But if people agrees on doing this, that can be much alleviated by defining macros to generate those types on the fly.

**Comment by equalsraf** [comment]
d52fb89fb85984df23b093637b9a8738fdcfe18a should take care of this one, Neovim now listens on a proper pipe by default
- [x]  fix listening server, Windows pipe names must always start be in the form "\\.\pipe\pipename"

**Comment by justinmk** [comment]
updated

**Comment by equalsraf** [comment]
FYI, (somehow :open_mouth:) my branch is now building luarocks from third-party - meaning it builds Neovim with just: 
- git, python2 (required by libuv)
- CMake
- Visual studio 2013 update 4

There is an automated build on [Appveyor](https://ci.appveyor.com/project/equalsraf/neovim)

**Comment by sethjackson** [comment]
@equalsraf Windows defines `SSIZE_T` in basetsd.h.
See: https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751%28v=vs.85%29.aspx

Whenever I've needed `ssize_t` I've just typedef'd it as `SSIZE_T`.

**Comment by equalsraf** [comment]
@sethjackson thanks. I was able to find ssize_t, what I couldnt find was SSIZE_MAX and ended up using [INT_MAX](https://msdn.microsoft.com/en-us/library/296az74e.aspx) instead. Maybe LONG_MAX might be the correct one depending on the architecture.

**Comment by fwalch** [comment]
> There is an automated build on Appveyor

Awesome! Maybe AppVeyor can already be enabled for `neovim/neovim` (cc @justinmk), and you can add the `appveyor.yml` to your Windows PRs. That way, the Appveyor build status will be visible in PRs, along with the Travis status.

**Comment by elmart** [comment]
> couldnt find SSIZE_MAX and ended up using INT_MAX instead

I think the way to go would be using preprocessor to define SSIZE_MAX where not present. It shouldn't be difficult, at first glance.

**Comment by sethjackson** [comment]
I agree with @elmart. I think defining `SSIZE_MAX` would probably be the right option.

`SSIZE_T` is a typedef for [LONG_PTR](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751%28v=vs.85%29.aspx) which is a `long` on 32-bit archs and `__int64` on 64-bit archs.

This should work:
```c
#if defined(_WIN64)
    #define SSIZE_MAX _I64_MAX
#else
    #define SSIZE_MAX LONG_MAX
#endif
```
wrapped in a Windows-specific header of course.

**Comment by equalsraf** [comment]
My quick fix at the time was [this](https://github.com/equalsraf/neovim/commit/69cf097b7069dc5c08c0c450160f86e130a32c0d)
```C
#ifndef SSIZE_MAX
# define SSIZE_MAX INT_MAX
#endif
```
Altough for full correctness it might be something like
```C
#ifndef SSIZE_MAX
# if defined(_WIN64)
// In 64bit ssize_t is a LONG_PTR(__int64)
#  define SSIZE_MAX I64_MAX
# else
// In 32bit ssize_t is a LONG_PTR(long)
#  define SSIZE_MAX LONG_MAX
# endif
// Maybe an else if  just in case we are not in MSVC
#endif
```

**Comment by equalsraf** [comment]
Ah beaten by 5s :D, @sethjackson that is it.

**Comment by equalsraf** [comment]
Here is a badge ![badge](https://ci.appveyor.com/api/projects/status/ki4bc2v81axm36ow/branch/tb-mingw?svg=true)

Configure AppVeyor to package the contents of the build folder [here](https://ci.appveyor.com/project/equalsraf/neovim/build/artifacts). 40mb is a bit, but it should include most things needed to debug. Linkage is static so it shouldn't need anything to run. If anyone gets a chance to run the binary in a windows machine let me know.

**Comment by ebraminio** [comment]
Yes, works.

**Comment by equalsraf** [comment]
@ebraminio :+1:

**Comment by equalsraf** [comment]
@justinmk cannot find any references to this anymore, I think it was fixed (#846? and maybe others)
- [x] we need a os_file_is_readable() in os/fs.c to fix eval.c:f_file_readable()

**Comment by equalsraf** [comment]
And now that the visual studio generator is working we can build using clang-cl (not for third-party deps though) just `cmake -G "Visual Studio 12" -T "LLVM-vs2013".

**Comment by aktau** [comment]
> Yes, works.

It *runs* on Windows?

Anyhow, the progress on this is unbelievably awesome (despite not being a Windows user).

Even if it just compiles, I'd like to advocate getting AppVeyor integrated like travis if at all possible, so we at least don't regress on compilation.

**Comment by ebraminio** [comment]
Yes.

![a](https://cloud.githubusercontent.com/assets/833473/6006519/57e1bf0c-ab28-11e4-967a-c7ed265f3898.png)

Of course the progress is awesome. Thank you guys!

**Comment by equalsraf** [comment]
@aktau sadly the mingw version in Travis seems too old, otherwise we could enable Mingw cross compile.

BTW the binaries run in wine, so if anyone wants to fire up the unit tests over TCP that would be cool (I have not tried the tests - but once I accidentally started nvim.exe against the python interpreter from Linux).

**Comment by equalsraf** [comment]
@justinmk a missing item that got lost along the way
- [ ] Implement unix_expandpath for Windows

**Comment by equalsraf** [comment]
Crash [dump](https://gist.githubusercontent.com/equalsraf/4db181c320d12b9a4998/raw/d7c76a31bd0be2739dc0ce8e91639c87992c5b18/gistfile1.txt) for MSVC [builds](https://ci.appveyor.com/project/equalsraf/neovim/build/53-tb-mingw/artifacts) - to reproduce connect over msgpack-rpc and then disconnect - I cannot debug this at the moment, since I have no workable debugger here. If anyone groks stack/register dumps, some insight would be appreciated. 

Looking at the trace it seems there might be something wrong with the **kv** in `call_set_error`. Oddly I cannot reproduce this with Mingw builds.

**Comment by equalsraf** [comment]
Nevermind, #1846 fixes it.

**Comment by aktau** [comment]
Seems to be GNU GAS (AT&T) syntax: http://en.wikibooks.org/wiki/X86_Assembly/GAS_Syntax

```
0x0061bd5b call_set_error+0x5b [c:\projects\neovim\src\nvim\msgpack_rpc\channel.c:728] in nvim: movl	0x0(%edx,%eax,4),%ecx

Is: 

Load 0x0(%edx,%eax,4) into register %ecx

Aka:

%ecx = *(%edx + (%eax * 4))

With

%eax = 00000000
%edx = 00000000

Gives

%ecx = *0x0 // NULL pointer dereference
```

Helpfully, the line in question was indicated:

```c
#define kv_A(v, i) ((v).a[(i)])
#define kv_size(v) ((v).n)

static void call_set_error(Channel *channel, char *msg)
{
  ELOG("Msgpack-RPC error: %s", msg);
  for (size_t i = 0; i < kv_size(channel->call_stack); i++) {
    ChannelCallFrame *frame = kv_A(channel->call_stack, i); // CRASH HERE
    frame->returned = true;
    frame->errored = true;
    frame->result = STRING_OBJ(cstr_to_string(msg));
  }

  close_channel(channel);
}
```

From this I surmise that %eax is the looping variable, and that the loop counter was **0** at the moment of crashing. We see that `kv_size()` returned at least **1**, else the loop wouldn't have been entered. The register containing it seems to have been overwritten as I can't see any other low-values in the register set.

It all seems a bit fishy, `channel->call_stack.n > 0` yet `channel->call_stack.a == NULL` (the array). That shouldn't happen. I can only think of two reasons why that would happen:

1) Badly initialized kvec containing garbage values, of which the second word is accidentally NULL. Yet this seems to be properly done by `kv_init()` in `register_channel()`. I believe the channel comes from `channel_from_stdio()` or `channel_from_stream()` because I don't see `job_out()` in the stacktrace.
2) Some part of the code actually manually setting `n` instead of going through the `kv_` functions
3) Wat...?

I would need to see the disassembly of the call_set_error() function that MSVC compiled to know a bit more. I'm also a bit less familiar with Microsofts 64-bit calling convention (or did you build a 32-bit binary? It kinda looks like it from your dump).

Notes to self:

```c
static void complete_call(msgpack_object *obj, Channel *channel)
{
  ChannelCallFrame *frame = kv_A(channel->call_stack,
                             kv_size(channel->call_stack) - 1);
  frame->returned = true;
  frame->errored = obj->via.array.ptr[2].type != MSGPACK_OBJECT_NIL;

  if (frame->errored) {
    msgpack_rpc_to_object(&obj->via.array.ptr[2], &frame->result);
  } else {
    msgpack_rpc_to_object(&obj->via.array.ptr[3], &frame->result);
  }
}
```

Are we sure the first line never underflows? I assume so, since the name of the function implies that there is at least one frame, but was just noting down.

Slightly related sidenote: why are we doing this with a `kvec` instead of a plain array? I suppose we need to set some upper limit to the stack frames and we could just make that part of the function: `ChannelCallFrame call_stack[MAX_CALL_FRAMES]` instead of `kvec_t(ChannelCallFrame *) call_stack`. It will likely get rid of some extra allocations and we'd have to worry less about these things.

@tarruda one callchain caught my eye. Consider a channel created with `channel_from_stream()`. It has `streams.uv != NULL`. `parse_msgpack()` gets called with `eof == 1`. So the following `if`-statement gets entered:

```c
  if (eof) {
    close_channel(channel);
    call_set_error(channel, "Channel was closed by the client");
    return;
  }
```

Let's look a bit closer at `close_channel()`:

```c
/// Close the channel streams/job and free the channel resources.
static void close_channel(Channel *channel)
{
  if (channel->closed) {
    return;
  }

  channel->closed = true;
  if (channel->is_job) {
    if (channel->data.job) {
      job_stop(channel->data.job);
    }
  } else {
    rstream_free(channel->data.streams.read);
    wstream_free(channel->data.streams.write);
    uv_handle_t *handle = (uv_handle_t *)channel->data.streams.uv;
    if (handle) {
      uv_close(handle, close_cb);
      free_channel(channel);
    } else {
      event_push((Event) { .handler = on_stdio_close }, false);
    }
  }
}
```

We see that near the end, if `streams.uv != NULL`, this will get run:

```c
    if (handle) {
      uv_close(handle, close_cb);
      free_channel(channel);
    } 
```

Incidentally, `free_channel()` frees the kvec (and everything else). **But** this will get used in `call_set_error()`.

Does this chain of events sound plausible?

In any case, I think we shouldn't conflate `close_channel()` and `free_channel()` in this way. Especially if it only happens very conditionally. Async programming is difficult enough without having to keep these things in one's head.

**Comment by aktau** [comment]
Alright... I hadn't noticed that #1846 fixes it. Though I'm still confused as to why this wasn't noticed before. If the chain of events I described above is what's happening, then we've gotten really lucky. But also, Valgrand or ASan would've/should've totally caught this. Why haven't they? Do we not test libuv stream based channels?

**Comment by tarruda** [comment]
Note about the terminal UI on windows: We may be able to get it working by simply using libuv tty handle(with #1820), turns out libuv implements an ansi escape code parser that translates into native windows console API calls! [ref](https://github.com/libuv/libuv/blob/v1.x/src/win/tty.c)

**Comment by equalsraf** [comment]
Most likely for the same reason I did not notice earlier, it only gets triggered if the client disconnects the channel and not if nvim closes first. In fact if we have a look at [coveralls](https://coveralls.io/builds/1872764/source?filename=src%2Fnvim%2Fmsgpack_rpc%2Fchannel.c#L728) that call never gets triggered.

#1846 was detected by coverity. The reason why I recalled it was because I introduced a similar [bug](https://github.com/neovim/neovim/commit/4e880f3f0058dce4a2a8d6fafeee2d89fad291d3) some time ago :S.

So your analysis above was spot on (now I kind of feel bad you had to do all that) 
- the size/capacity values were most likely MAX - this is as far as I got by debugging, before assuming an extra free was around
- yes I'm doing 32bit builds since I'm never sure where I will end up running them, I still have to check how to enable 64 bit builds in cmake+msbuild

**Comment by equalsraf** [comment]
@tarruda :+1:

**Comment by elmart** [comment]
> Incidentally, free_channel() frees the kvec (and everything else). But this will get used in call_set_error().

Note that coverity caught a couple of errors related to this (coverity/68484, coverity/68485).
I will be soon trying to fix them following @tarruda's instructions [here](https://github.com/neovim/neovim/issues/1938#issuecomment-72906671).

**Comment by aktau** [comment]
> Note that coverity caught a couple of errors related to this (coverity/68484, coverity/68485).
I will be soon trying to fix them following @tarruda's instructions here.

Great. But this module needs to be refactored too, as @tarruda has alluded to :). While we're add it we can also unify the function namespace to `chan_`.

**Comment by aktau** [comment]
> @aktau Most likely for the same reason I did not notice earlier, it only gets triggered if the client disconnects the channel and not if nvim closes first. In fact if we have a look at coveralls that call never gets triggered.

That coveralls stuff is pretty awesome now that it works. We could've caught it...

> So your analysis above was spot on (now I kind of feel bad you had to do all that)

Please don't be :). I understand a tiny bit more of our codebase and that's always nice. Plus I'm glad to be able to help even I don't really have the time or equipment at times to contribute code. I'm not a Windows user but I understand how important it is to have a nice, cross-platform codebase. Windows users deserve a great editor too. So thanks for working on it!

**Comment by sethjackson** [comment]
@equalsraf to do 64-bit builds with MSBuild you can do the following:

```
$ cd .\build

$ cmake -G "Visual Sutdio 12 2013 Win64" ..

$ cmake --build . --config Release
```

You can pass any config types the project has enabled to the `--config` flag.

**Comment by equalsraf** [comment]
@sethjackson thanks, it works for Neovim. third-party is trickier, libuv was still being built for 32bit - pushing a patch now

**Comment by equalsraf** [comment]
@justinmk 
- [ ] bring back utf16_to_enc()/enc_to_utf16() - needed to convert string encoding when calling windows functions

**Comment by justinmk** [comment]
@equalsraf  updated. still not sure about os_file_is_readable though

**Comment by syswow** [comment]
@equalsraf  would you please export the appveyor.yml, i cannot finish the last step (Uploading artifact Neovim.zip).

**Comment by equalsraf** [comment]
@syswow [here](https://gist.github.com/equalsraf/e42f0aa70326485553e1) you go, hope it helps

**Comment by syswow** [comment]
@equalsraf  Great, it works, thanks.

**Comment by equalsraf** [comment]
After the last rebase we can now start some real testing with the UI.

Sadly my own GUI implementation is too outdated to work properly and there is some issue with the python-client GUI in Windows. Still it worked well enough to grab this screenshot.

![account1](https://cloud.githubusercontent.com/assets/715238/6265127/95aef992-b822-11e4-8c6a-4e789ef20594.png)

(running in Wine, sorry about the colors :S, the UI doesnt render text properly with the default color scheme)

**Comment by tarruda** [comment]
@equalsraf did you try [neovim.as](https://github.com/coolwanglu/neovim.as)?

**Comment by jasonwilliams200OK** [comment]
> sadly the mingw version in Travis seems too old, otherwise we could enable Mingw cross compile.

FWIW, MinGW is also installed on AppVeyor at `C:\MinGW`.
Ref: http://www.appveyor.com/docs/installed-software.
I think you can invoke it as:

```bash
# not sure about this
sh: /do/whatever
```

same as you can invoke PowerShell:

```powershell
# since posh doesn't mind forward-slashes (and back-slashes)
ps: /do/whatever
```

The default shell is cmd, you could do something like

```cmd
cmd: c:\MinGW\msys\1.0\bin\sh.exe /do/whatever
:: or
c:\MinGW\msys\1.0\bin\sh.exe /do/whatever
```

**Comment by equalsraf** [comment]
@jasonwilliams200OK  I got around it by downloading a binary mingw toolchain into Travis, it now cross compiles to mingw too.

Native builds using msys2 dont work right now, the third-party CMakelists.txt needs some fixing and don't have a windows machine on me right now. I'll have a look at enabling it in appveyor.

**Comment by jasonwilliams200OK** [comment]
@equalsraf great! Here is an example of appveyor.yml where they are installing custom version of mingw: [sass/libsass/appveyor.yml](https://github.com/sass/libsass/blob/baaf1ccf2d75c7a94abd2a48391f128278930be5/appveyor.yml#L27-L37).

**Comment by equalsraf** [comment]
Back, sort of. Trying to rebase against master, almost through, but I'm bumping into some changes:

- [ ] terminal emulator and libvterm - I feel tempted to disable this in Windows, IIRC libvterm requires glib-2.0
- [ ] pty_process.c seems UNIX specific (termios.h)
- [ ] job.c, waitpid(-1, ...) equivalent for Windows

**Comment by tarruda** [comment]
>  terminal emulator and libvterm - I feel tempted to disable this in Windows, IIRC libvterm requires glib-2.0

libvterm has no dependencies. It's pure C99 so maybe microsoft compiler will have trouble.

> pty_process.c seems UNIX specific (termios.h)

For now pipe_process.c can be used(disable pty_process.c compilation and #define pty_* functions as pipe_*. Later I will try to implement a winpty_process.c on top of [winpty](https://github.com/rprichard/winpty)

> job.c, waitpid(-1, ...) equivalent for Windows

SIGCHLD handling can simply be removed(through #ifdef?), it only exists to support processes spawned by pty_process.

**Comment by elyscape** [comment]
> Later I will try to implement a winpty_process.c on top of [winpty](https://github.com/rprichard/winpty)

That might be a mistake.  winpty introduces latency and doesn't give you a raw bytestream.

**Comment by equalsraf** [comment]
> libvterm has no dependencies. It's pure C99 so maybe microsoft compiler will have trouble.

For some reason I thought it depended on glib, maybe I checked the wrong libvterm - anyway I a have recipe for building using CMake. It still needs perl to generate code, but maybe that can be solved by including the generated source in the libvterm repo.

> For now pipe_process.c can be used(disable pty_process.c compilation and #define pty_* functions as pipe_*. 

This works for now, minus the pty_process close master.

```
# define pty_process_init(job) pipe_process_init(job)
# define pty_process_spawn(job) pipe_process_spawn(job)
# define pty_process_destroy(job) pipe_process_destroy(job)
# define pty_process_close(job) pipe_process_close(job)
// FIXME: this is not right
# define pty_process_close_master(job) pipe_process_close(job)
# define pty_process_resize(job, width, height)
```

> SIGCHLD handling can simply be removed(through #ifdef?), it only exists to support processes spawned by pty_process.

Ok on this front too, just guarded those blocks with a define FEAT_PTY_PROCESS.

With libvterm building, I've got the Mingw builds going again in travis, but test49 seems to be blocking.

The MSVC builds will have to wait a bit longer until I get my hands on a Windows machine.

**Comment by tarruda** [comment]
> That might be a mistake. winpty introduces latency and doesn't give you a raw bytestream.

Do you have an alternative? Windows has nothing close to the master/slave terminal API, so  console programs must be spawned in builtin console window. In fact, a quick research suggests that custom terminal emulators such as [Console2](http://sourceforge.net/projects/console/) are implemented using the same approach as winpty(hidden console window with polling through the win32 API)

**Comment by elyscape** [comment]
Yeah, I suppose that's a fair point, and I unfortunately don't have an alternative other than just using pipes, which introduces its own issues.

**Comment by equalsraf** [comment]
> libvterm has no dependencies. It's pure C99 so maybe microsoft compiler will have trouble.

It throws about a dozen errors with Variable Length Arrays, and an unsupported escape sequence \e.

@tarruda I ended up adding a def FEAT_TERMINAL to enable/disable support for the terminal emulator in  	8365c93. 

AppVeyor builds are back up :D

**Comment by equalsraf** [comment]
- [ ] busted is disabled for Windows because of lubyk/xml#2, the fix is merged into master though

**Comment by elyscape** [comment]
> an unsupported escape sequence \e

`\e` isn't supported by MSVC.  If we replace it with the literal `\x1B`, it should work fine.

**Comment by elyscape** [comment]
> It throws about a dozen errors with Variable Length Arrays

This is still a problem, though.  A possible, albeit suboptimal, fix might be to patch the code that uses VLAs to instead allocate on the heap with `new whatever[len]`.

**Comment by tarruda** [comment]
> It throws about a dozen errors with Variable Length Arrays, and an unsupported escape sequence \e.

@leonerd would it be ok to remove \e and VLAs from libvterm? It would allow compilation with visual studio.

**Comment by leonerd** [comment]
Replacing the `\e` with `\x1B` should be simple enough.

As to the VLAs, I did spend a while staring at those some time ago and came to the conclusion that a few of them at least are nontrivial to hack at. I'll take another look and see what I can manage.

**Comment by elyscape** [comment]
@leonerd It should be possible to replace them with [`_alloca()`](https://msdn.microsoft.com/en-us/library/wb1s57t5.aspx).  Unfortunately, on basically everything that isn't MSVC, the function's called `alloca()`.  But that's easily worked around by just using `alloca()` and defining it around for MSVC:
```c
#ifndef alloca
#define alloca(size) _alloca(size)
#endif
```

**Comment by dusans** [comment]
Just as a note. I switched from Console2 to ConEmu. 

ConEmu inject a dll: http://conemu.github.io/en/ConEmuHk.html
> ConEmu need to intercept large amount of Windows API functions, otherwise it is not possible neither to avoid certain MicrosoftBugs nor implement much of juicy features.

**Comment by tarruda** [comment]
@dusans Console2 also injects a dll, I've seen in the code. They spawn the child process in paused state, inject some machine code to load the dll and create a thread to run the injected machine code.

I've read that injecting a dll is more reliable because polling the console window can result in lost data, but I've also read that in practice this never happens

**Comment by fwalch** [comment]
@equalsraf Do you think we can start merging some of your changes into masters as small, individual PRs? Things that come to mind are a basic `appveyor.yml` (only enabled for `nightly` tag for now, to build once per day) and fixes for building the dependencies (to put them into `neovim/deps` with `bot-ci`). And then continue from there... what do you think?

**Comment by jasonwilliams200OK** [comment]
IMO: it will make sense to raise min required version here for VS to 2015, which is free for OSS projects: http://blogs.msdn.com/b/visualstudio/archive/2015/03/31/announcing-the-visual-studio-2015-product-line.aspx. The preview has many C99 features which VS2013 is missing. Also, C99 will be fully supported by the time 2015 will reach RTM.
Basically from cl.exe's perspective, the higher version number means more language features, regardless of those preview/beta/rc tags attached to VS.

**Comment by equalsraf** [comment]
@fwalch that is the general idea. I usually rebase #810 to keep the more stable commits at the start. Other than that I just prioritize based on what changes are likely to cause a conflict on the next rebase :D

- Next on my list is probably 554c62a464f4d7cce2cc1d47aefe3d406d7ba554, because it touches way too many files.
- Feel free to scanvenge #810 for any useful commit, e.g. c6f54071e143bea7368e7e2b76677a702e3d7283, a8651c006c1b18917138186eb21c5fd58bb578a1 are probably useful for non-Windows.

I like your idea about the dependencies It is probably possible to get a working PR that builds them for Mingw in Travis and MSVC in Appveyor.

@jasonwilliams200OK I agree, the only reason I have not moved to VS2015 is because of libuv (gyp did not work with it - https://code.google.com/p/gyp/issues/detail?id=446). :smiley: my Windows machine is a mess I have so many dev environments mixed in here, its a small miracle anything works at all.

**Comment by jasonwilliams200OK** [comment]
@equalsraf, here is a commit which brings VS2015 support to gyp https://code.google.com/p/gyp/source/detail?r=2013 (4 months old), corresponding mirror commit: https://github.com/svn2github/gyp/commit/21754fa1ffb12acc4782ad345055b6ad5bae79b0. Not sure if this fixes the issue to get the compiler working. 

Also going by the progress here: http://code.google.com/p/chromium/issues/detail?id=440500, (scroll till bottom) they have removed a lot of stuff to consume  implementations (such as snprintf) in their code base. I think until VS2015 gets to RTM, gyp will be fine-tuned accordingly. 

Related: https://github.com/TooTallNate/node-gyp/issues/555.

For experimentation, I usually use Vagrant (http://vagrantup.com/) to establish the disposable environment (win8.1 box: https://atlas.hashicorp.com/fundpuls/boxes/windows-81). :)

**Comment by equalsraf** [comment]
@jasonwilliams200OK  Impeccable timing, libuv also has a snprintf replacement (libuv/libuv#329) in there :D

I was aware of the gyp commit, but I don't think it is working for libuv yet. I'll give it another shot when I get the chance.

FYI I'm getting a network timeout when downloading the vagrant box.

**Comment by jasonwilliams200OK** [comment]
@equalsraf, it seems like that provider's server is down. I tested with official IE VM provider for testers (http://modern.ie), and it worked!

```cmd
:: install vagrant from http://vagrantup.com/
:: install virtualbox from Oracle
vagrant init http://aka.ms/vagrant-win81-ie11
vagrant up
:: install VS in VM
```

BTW it is just announced that some new edition of VS called VS Code has been previewed for Linux, Mac and Windows: https://www.visualstudio.com/products/code-vs (me looking forward to get vim extension working on it: https://github.com/jaredpar/VsVim) :smiley:

**Comment by elyscape** [comment]
> (me looking forward to get vim extension working on it: https://github.com/jaredpar/VsVim)

Visual Studio Code is based on Electron, aka Atom Shell, so you'll have more luck basing the plugin on https://github.com/carlosdcastillo/vim-mode.

**Comment by chapgaga** [comment]
@equalsraf why vs 2013 can be used to build neovim? i think neovim uses lot's of linux api.

**Comment by equalsraf** [comment]
@chapgaga for the *majority* of things Neovim uses libuv, which works both in Windows and Unix systems. Also a lot of the POSIX functions are also supported by the Microsoft compiler.

There are some things that still need to be ported to work on Windows like if_cscope, or the terminal.

**Comment by jasonwilliams200OK** [comment]
Here is the list of C99/C11/C++11/C++14/C++17 features supported in VS2015 RC: http://blogs.msdn.com/b/vcblog/archive/2015/04/29/c-11-14-17-features-in-vs-2015-rc.aspx (see the FAQs).

From comments, STL (person) stated:

> I personally consider 2015 RC's compiler/CRT/STL to be *dramatically* higher quality than 2013 Update 4's. The new features are nice, but we've also fixed so many bugs in the existing features. The STL alone has received something like 200 fixes (I'd have to re-count the fixes between 2013 and 2015 CTP1 that I blogged about; there have been almost 100 between CTP1 and RC), and we've been very careful to avoid introducing regressions. (I do have a fix for one regression that my <functional> overhaul introduced in RC, where bind()'s newly overloaded function call operator could fail to compile; that will be checked in along with constexpr.)

**Comment by jasonwilliams200OK** [comment]
@equalsraf, FYI libuv [v1.6.0](https://github.com/libuv/libuv/releases/tag/v1.6.0) with your fix (latest patched version is v1.6.1). Now, the official libuv release builds with VS2015. :)

Aside: to experiment with the latest/unreleased builds of VC compiler, you can try their web-based compiler http://webcompiler.cloudapp.net/ (ref: http://blogs.msdn.com/b/vcblog/archive/2015/06/02/constexpr-complete-for-vs-2015-rtm-c-11-compiler-c-17-stl.aspx)

**Comment by equalsraf** [comment]
@jasonwilliams200OK thanks for the heads up

Aside: chocolatey packages for Neovim and my qt gui
```
choco sources add -source https://ci.appveyor.com/nuget/equalsraf-neovim -name equalsraf-neovim
choco install -pre neovim
choco sources add -source https://www.myget.org/F/equalsraf/ -name equalsraf
choco install -pre neovim-qt
```

**Comment by fwalch** [comment]
@equalsraf Whoa, awesome! Maybe add it to the install instruction page on the wiki? :-)

**Comment by jasonwilliams200OK** [comment]
@equalsraf, :+1: :100: this is a brilliant news! and I totally agree with @fwalch to spread the word.

I just installed neovim using chocolatey, on running `nvim test.txt`, I get the following warning in PowerShell:

> Terminal UI was disabled at compile time, Neovim will not accept input from the console

Aside, is there any remaining showstopper in building neovim with VS2013 / VS2015?

**Comment by equalsraf** [comment]
@jasonwilliams200OK the terminal UI still doesnt work, if you run nvim-qt you should get a running GUI.

> Aside, is there any remaining showstopper in building neovim with VS2013 / VS2015?

Minor things I hope, I'll find a Windows machine in a bit. I'll have to test in Appveyor to check the automated builds.

**Comment by jasonwilliams200OK** [comment]
Thanks! :+1: 

Neovim running on Windows! :tada: :balloon: 

> I'll find a Windows machine in a bit

The Docker project (https://github.com/docker/docker) is getting official support from Windows team. So at some point in future, we will be able to run CLI-only/headless windows (analogous to Windows Server Core) docker. Also, VC team is considering to isolate VC compiler from VS. With that said, chances are: in future, we may be able to have a very light-weight Docker image with predefined steps to build the native binaries using cl.exe. Another brilliant thing happened is that Clang has been brought to Windows: http://blogs.msdn.com/b/vcblog/archive/2015/05/01/bringing-clang-to-windows.aspx (corresponds to https://github.com/microsoft/clang).

**Comment by equalsraf** [comment]
> Neovim running on Windows! :tada: :balloon:

Still ways to go - LOL it just dawned on me that #810 was opened in June 2014.

In a nutshell:
- MSVC 2015 builds using libuv 1.6
- MSVC 2013, only builds if we dont use VLAs (#810 has a commit for this), for now I'll leave it since lots of people still use 2013

Good news on the Docker and VC fronts :D. I tried the Clang-cl windows builds earlier this year but they were not working for me, CMake got confused about the toolchain and mixed in compiler flags.

> @equalsraf Whoa, awesome! Maybe add it to the install instruction page on the wiki? :-)

I'm really not sure how stable this is. I haven't even looked over things relating to encoding, i.e. needs lots of work. But at least it should help people give it a spin - #810 has a bundle of instruction for all the systems at the top.

**Comment by jasonwilliams200OK** [comment]
[neo-build strategy for neovim]

@equalsraf, by the time noevim will get its first release, VS2015 will be RTM'd (few weeks before Windows 10 release, which is scheduled to release on July 29).

With that said, I have two suggestions (ahem): 

1. Declare VS2015 as minimum requirement to build neovim. (one of the charms of semantic versioning is that you can make breaking changes and such decisions when major-version == 0, after that breaking change(s) require major version bump)
2. Remove all the guarded inclusions and C99 fallback implementations for VS < v2015.

Since there is full featured community version and win32 desktop only express versions available for free of cost, the Windows users will have enough choices: MinGW, VS2015 pro/ultimate/express-for-desktop/community to build neovim source.

Besides, you guys will probably make the win32 installer available for mainstream consumer. Building from scratch is just for enthusiasts and contributors, so I don't see it as a real deal-breaker.

Note that two main Windows supporting CI systems: AppVeyor and Jenkins (https://github.com/nodejs/io.js/pull/478#issuecomment-111682322) already support VS2015 RC. :)

**Comment by fwalch** [comment]
@jasonwilliams200OK That sounds like a very good plan IMO :+1:

While installing VS 2015 [requires at least Windows 7](https://www.visualstudio.com/visual-studio-2015-system-requirements-vs#VSC_RC), it can [build for targets as far back as Windows XP](https://www.visualstudio.com/en-us/products/visual-studio-2015-compatibility-vs.aspx) (for the RC, at least). So releasing an installer for all relevant Windows versions should be no problem for us.

**Comment by equalsraf** [comment]
@jasonwilliams200OK those are my thoughts exactly :+1:

**Comment by leonardinius** [comment]
I would like to stop by and thank you all for your hard work on Windows support. I'm using nvim-qt lately as my default GIT_EDITOR and could not praise it enough.

Kudos! Highly appreciated.

**Comment by justinmk** [comment]
Updated the issue description after reviewing all Windows-related discussions. @equalsraf some questions if you can:

> busted is disabled for Windows because of lubyk/xml#2, the fix is merged into master though

Anything we need to do here?

> With libvterm building, I've got the Mingw builds going again in travis, but test49 seems to be blocking.

Is test49 still an issue?

Where are we at with libintl?

**Comment by equalsraf** [comment]
Hi @justinmk I've been away for a *bit* and don't have windows machines to test until next week.

- For busted tests - xml should work for MSVC but not for Mingw (lubyk/xml#4), I'm pretty sure the lua nvim-client does not build for windows though.
- Not sure about test49. #810 still does not implement :terminal anyway and I never pushed the changes to libvterm
- No changes on libintl
- At this point might as well change the issue title to VS2015

**Comment by justinmk** [comment]
Updated. Thanks @equalsraf

**Comment by bfredl** [comment]
I was trying to wrap my head around the locale `&encoding` situation on windows. I found this little piece was removed from `enc_locale()` even in the root commit:

```
#ifdef WIN3264
    long	acp = GetACP();

    if (acp == 1200)
	STRCPY(buf, "ucs-2le");
    else if (acp == 1252)	    /* cp1252 is used as latin1 */
	STRCPY(buf, "latin1");
    else
	sprintf(buf, "cp%ld", acp);
#else
```

In practice this means that `&encoding` will use the default utf-8, which should be the right thing to do (IIRC the "code page" on windows can _never_ be utf-8, e.g. on a swedish windows install the default will always be `encoding=latin-1` in gvim.exe). But I wonder if we still should retain this information, e.g. so that `:set fenc[s]=default` uses the locale code-page just like gvim.exe

**Comment by justinmk** [comment]
I suspect most windows users do not at all want the 8 bit codepage as the default. Its existence is an accident of the windows cmd.exe legacy. We should default to utf8, like sublime and atom.

**Comment by bfredl** [comment]
@justinmk Definitely. My question was more about `&fencs`, whenever `enc=utf-8` (or synonym), we have per default

    fileencodings=ucs-bom,utf-8,default,latin1

So whenever a file _fails_ to load as `utf-8` do we fallback to the system code-page (which not necessarily is what `cmd.exe` is using, last time I checked  "MS-DOS Latin" was still the default ) or do we always use `latin-1` ? I'm fine either way really, just want to be sure we make an explicit decision (and not just what POSIX-minded code happened to do on windows, even if it accidentally is the right thing :) )

**Comment by justinmk** [comment]
@bfredl I misunderstood then. :+1: to restoring that logic.

**Comment by jasonwilliams200OK** [comment]
Regarding:
>  Add include guards: sys/time.h unistd.h

With MSVC, we can probably use `<io.h>` for `<unistd.h>` and `<pdh.h>` for `<sys/time.h>` (I haven't used `<pdh.h>` myself, just found folks use it to get low-level processor time info etc. on windows, related: https://github.com/libuv/libuv/issues/342).

**Comment by justinmk** [comment]
@jasonwilliams200OK Thanks! I see #810 is using `<io.h>`, and the only place left that we are using `sys/time.h` is in `log.h`, for `strftime()`. It seems that MSVC's `time.h` provides `strftime()`, which #810 also has included. So I'm marking that item completed. @equalsraf can you confirm?

**Comment by justinmk** [comment]
I just removed both remaining instances of `sys/time.h`, and it compiled fine.

**Comment by christiand** [comment]
Not sure if this is the right place to post, but I'm curious as to the team's thoughts regarding a Windows GUI. No so much "when", but in what form? i.e. if i'm looking to use it professionally and contribute, where to best spend the energy? The assorted "nvim-inside-3rdparty-app" like atom/qt/etc. are interesting.. but ultimately solo interest projects at this point. I'd like to believe that there would be a native supported experience a la gvim.
I'm not to familiar with where you guys drew the line when refactoring vim (and I don't know vim source to begin with..) but is gvim's gui layer a candidate at all?

**Comment by equalsraf** [comment]
@christiand there is a an issue for discussing a reference GUI in Neovim (#1639) you might want to have a look at.

For windows specifically there are a number of third party clients that work right now (qt, atom, C#), there also the one implemented as part of neovim/python-client. Technically Neovim treats all UIs as clients, even the built in UI - so all of them get the same API, at the very least this should make it easier for the best solution to race to the top.

**Comment by equalsraf** [comment]
**Question**: How do we feel about static vs dynamic linking and distributing DLLs?

For Unix Neovim links statically as much as possible but for Windows this makes things harder:
- trying to find pre-compiled third party dependencies (e.g. libintl or iconv - seems nobody distributes /MT libraries for MSVC).
- Some of our pending issues with lua relate to using static linking (luarocks dislikes this in rocks recipes)
- Our third-party recipes get a bit more complex in order to enforce /MT

Using dynamic linking also has its own problems:
- Finding and bundling DLL dependencies with binaries (cmake has BundleUtilities, but I never tested) - there are scripts for this lying around the internet
- Need to bundle the MSVC runtime DLLs too

**Comment by jasonwilliams200OK** [comment]
One swell approach could be to first clone them conditionally in CMake (quasi-submodule):

```cmake
if(WIN32)
  # first delete the existing
  execute_process( COMMAND powershell "rm -r -fo ${CMAKE_CURRENT_LIST_DIR}/libint 2>&1> $null" ) 

  # then clone
  execute_process( COMMAND git clone https://github.com/evaleev/libint --recursive -b v1 ${CMAKE_CURRENT_LIST_DIR}/libint ) 
end
```

Then build them. However, there are at least two issues with this approach:

* If the downstream doesn't have CMake, then we will need to configure that in our script too (perhaps add a separate cmake module `.cmake` for each lib).
* If the dependency lib doesn't exist on github/git (using SVN/CVS etc.), then we will need to mirror it.

Besides technical issues, I am not sure how does the license [in]compatibilities work in [up/down]stream / submodule scenarios, but it would be good to iron that out with some OSS solicitor too. :)

This is extra work at build time, but I think this way we will have control and freedom to however we want to ship it (we may end up building `.dll` and bundle it in installer, instead of statically linking `.lib` for some good reasons or vice versa).

**Comment by equalsraf** [comment]
> One swell approach could be to first clone them conditionally in CMake (quasi-submodule):

This we already have, but we grab tarballs of specific versions instead of cloning repositories.

> - If the downstream doesn't have CMake, then we will need to configure that in our script too

This is what the modules in third-party/cmake/Build*.cmake are doing right now.

But some cases are trickier - e.g. iconv only builds with autotools and these are only available in Unix or MSYS and some compile steps are non trivial (merge config.h.in with config options), nothing short of writing our own CMake recipe would work. The same probably applies for libintl and other GNU libraries.

> f the dependency lib doesn't exist on github/git (using SVN/CVS etc.), then we will need to mirror it.

Yup, this is what is done for libtermkey and unibilium.

> Besides technical issues, I am not sure how does the license [in]compatibilities work in [up/down]stream / submodule scenarios, but it would be good to iron that out with some OSS solicitor too. :)

I certainly hope not, since our dependencies are the same for Windows and Unix (so far) - but it would not hurt to check.

My main doubts in regards to MSVC is whether or not to build static libraries (/MT) of our dependencies, or to just build DLLs (and tweak the build process to bundle them with the Neovim binary).

As for grabbing third-party builds from someplace else - we will have to do it, at least for some runtime tools  (#1811) but maybe not for build dependencies.

**Comment by equalsraf** [comment]
- [ ] Get ffi working with MSVC - specially how to handle __pragma (see d530fb62dd52a0724fab38be49514a47464eef75 for more details)

**Comment by justinmk** [comment]
@equalsraf nice work. But the functional tests don't use FFI, do they run?

**Comment by equalsraf** [comment]
They don't run, the lua client was not working.in Windows last I checked. I have some fixes lying around, but there were some issues with luarocks failing to build.

**Comment by dt1973** [comment]
Would it be possible to come up with some sorta Windows installer?

**Comment by fwalch** [comment]
@dt1973 I think it is already being addressed in #810 (https://github.com/equalsraf/neovim/commit/025760f573af22f5a552d89de747761019f52868), which is merged into this repo step-by-step.

**Comment by equalsraf** [comment]
@fwalch, yes I added an installer its available as an artifact in appveyor. For some reason the $PATH is not being set, seems to be a problem with CPack/NSIS.

**Comment by equalsraf** [comment]
@leonerd we have a working libvterm for Windows with https://github.com/equalsraf/libvterm-win/commit/59ef3effbd66038ae467c2013a3244cfadca9273 builds with Mingw/MSVC.

Thanks @brcolow

**Comment by brcolow** [comment]
Also builds with cygwin (which is what I used) (GNU Make 4.1 Built for x86_64-unknown-cygwin)

**Comment by equalsraf** [comment]
The changes to STATIC_ASSERT in https://github.com/neovim/neovim/pull/4131/commits/6a35f2ac8bd7a3a0e8c9b988b85237424bc10e73 don't build on MSVC

```
eval.c(12962): warning C4003: not enough actual parameters for macro 'STATIC_ASSERT_STATEMENT'
eval.c(12962): error C2059: syntax error: ')'
```

I was expecting __VA_ARGS__ to actually work here but it seems to expand with a trailing coma. this

```
  STATIC_ASSERT(sizeof(u.prof) == sizeof(u) && sizeof(u.split) == sizeof(u),
      "type punning will produce incorrect results on this platform");
```

turns into

```
  do {  static_assert(sizeof(u.prof) == sizeof(u) && sizeof(u.split) == sizeof(u),
    "type punning will produce incorrect results on this platform", );  } while (0);
```

@ZyX-I any particular reason for STATIC_ASSERT to have variable arguments? all cases seem to take two arguments (cond, msg)

**Comment by jasonwilliams200OK** [comment]
@equalsraf, could you provide steps to repro? I tested the following code with VS2015 update 2:

```c
struct test
{
    int prof;
    int split;
};

#define STATIC_ASSERT_PRAGMA_START
#define STATIC_ASSERT_PRAGMA_END
#define STATIC_ASSERT(cond, msg) \
    do \
    { \
      STATIC_ASSERT_PRAGMA_START \
      STATIC_ASSERT_STATEMENT(cond, msg); \
      STATIC_ASSERT_PRAGMA_END \
    } while (0)

# define STATIC_ASSERT_STATEMENT(cond, msg) static_assert(cond, msg)

int main(void)
{
    struct test u;
    STATIC_ASSERT(sizeof(u.prof) == sizeof(u) && sizeof(u.split) == sizeof(u),
      "type punning will produce incorrect results on this platform");
}
```

via http://webcompiler.cloudapp.net/ using `/TC` <sub>([both capital letters](https://msdn.microsoft.com/en-us/library/032xwy55.aspx))</sub> compiler flag, and it gives `type punning...` (expected) compiler error.

**Comment by equalsraf** [comment]
@jasonwilliams200OK sadly no, your example works fine here.

**Comment by equalsraf** [comment]
This seems like a match http://stackoverflow.com/questions/10002296/macroized-parameters and the proposed fix seems to work, or at least builds, but I don't have a fix that works in all compilers.

**Comment by jasonwilliams200OK** [comment]
@equalsraf, my example fails if we use `#define STATIC_ASSERT(...)` and then `STATIC_ASSERT_STATEMENT(__VA_ARGS__)`.

Found related active bugs:
https://connect.microsoft.com/VisualStudio/Feedback/Details/1232378 and
https://connect.microsoft.com/VisualStudio/Feedback/Details/1099052.

**Comment by justinmk** [comment]
Replaced by https://github.com/neovim/neovim/issues/5229
